Feature/lazy collections beta

hg-ms opened the pull request:
Lazy collections implementation
-LazyArrayList
-LazyHashMap
-LazyHashSet

hg-ms committed:
Lazy collections implementation

Lazy collections implementation
-LazyArrayList
-LazyHashMap
-LazyHashSet

hg-ms committed:
Update Jenkins config to match master

fh-ms commented on base/src/main/java/one/microstream/collections/lazy/LazyCollection.java:
```diff
@@ -0,0 +1,116 @@
+	public default Stream<E> parallelStream()
+	{
+		return Collection.super.parallelStream()
+			//.onClose(Static.createLazyStreamCloseHandler(this))
```
No close-handler in parallel stream?

hg-ms replied:
forgot to enable it -> enabled

fh-ms commented on base/src/main/java/one/microstream/collections/lazy/LazyArrayList.java:
```diff
@@ -0,0 +1,1367 @@
+ *
+ * @param <E> the type of elements in this collection
+ */
+public class LazyArrayList<E> extends AbstractList<E> implements LazyList<E>, RandomAccess
```
Making all lazy implementations 'final' is probably a good idea.

hg-ms replied:
agreed, done

fh-ms commented on base/src/main/java/one/microstream/collections/lazy/LazyArrayList.java:
```diff
@@ -0,0 +1,1367 @@
+	 * 
+	 * @return the current number of internal segments.
+	 */
+    public long getSegmentCount()
```
format

hg-ms replied:
caused by spaces instead of tabs, corrected

fh-ms commented on base/src/main/java/one/microstream/collections/lazy/LazyArrayList.java:
```diff
@@ -0,0 +1,1367 @@
+     * 
+     * @return an Iterable over the Segment in this list
+     */
+	public Iterable<? extends Segment> segments()
```
format

hg-ms replied:
caused by spaces instead of tabs, corrected

fh-ms commented on base/src/main/java/one/microstream/collections/lazy/LazyArrayList.java:
```diff
@@ -0,0 +1,1367 @@
+	public int lastIndexOf(final Object element)
+	{
+        return this.lastIndexOfRange(element, 0, this.size);
+    }
```
format

hg-ms replied:
caused by spaces instead of tabs, corrected

fh-ms commented on base/src/main/java/one/microstream/reference/LazyClearObserver.java:
```diff
@@ -0,0 +1,35 @@
+ * Classes that implement that interface can utilize {@link one.microstream.reference.ObservedLazyReference.Default}
+ * to gain control of the unloading of {@link Lazy} references.
+ */
+public interface LazyClearObserver
```
That should be a Controller, not an Observer

hg-ms replied:
ok, renamed

fh-ms commented on base/src/main/java/one/microstream/reference/ObservedLazyReference.java:
```diff
@@ -0,0 +1,107 @@
+ *
+ * @param <T> type of the lazy loaded object
+ */
+public interface ObservedLazyReference<T> extends Lazy<T>
```
See LazyClearObserver

hg-ms replied:
ok, renamed

fh-ms commented on persistence/binary/src/main/java/module-info.java:
```diff
@@ -40,5 +40,6 @@
 	exports one.microstream.persistence.binary.one.microstream.collections;
 	
```
Add new packages as exports

hg-ms replied:
ok, done

fh-ms reviewed (commented):
I added some 'format' comments, and then I realized that there are a lot of places where the indentation is odd, but maybe it is just the GitHub viewer.

hg-ms committed:
fixed review findings

- fixed whitespaces
- renamed ObservedLazyReference to ControlledLazyReference
- renamed BinaryHandlerLazyObservable  to BinaryHandlerControlledLazy
- renamed LazyClearObserver to LazyClearController
- made lazy collections implementations final
- made LazySegmentUnloader implementations final
- made binary handler final

hg-ms committed:
added collections.lazy to exports

fh-ms commented on base/src/main/java/one/microstream/reference/ControlledLazyReference.java:
```diff
@@ -0,0 +1,108 @@
+		* Standard constructor used by normal logic to instantiate a reference.
+		*
+		* @param subject the subject to be referenced.
+		* @param lazyClearController the lazy reference clear observer.
```
typo: observer -> controller

hg-ms replied:
fixed

fh-ms reviewed (commented):
One last typo.

hg-ms committed:
renabled parallelStream close handler

hg-ms committed:
one more type fixed
